kv/kvserver: TestFlowControlRaftSnapshotV2 failed

kv/kvserver.TestFlowControlRaftSnapshotV2 [failed](https://mesolite.cluster.engflow.com/invocations/default/411f3270-ac0d-42ca-bfc5-c4d92022f3d6?testReportRun=1&testReportShard=23&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [13b35125a8626a1ea6f11f9d3654440e74e8055b](https://github.com/cockroachdb/cockroach/commits/13b35125a8626a1ea6f11f9d3654440e74e8055b):


```
=== RUN   TestFlowControlRaftSnapshotV2
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestFlowControlRaftSnapshotV22841798680
    test_log_scope.go:76: use -show-logs to present logs inline
    flow_control_integration_test.go:3088: -- test log scope end --
test logs left over in: outputs.zip/logTestFlowControlRaftSnapshotV22841798680
--- FAIL: TestFlowControlRaftSnapshotV2 (40.49s)
=== RUN   TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2/kvadmission.flow_control.mode=apply_to_elastic
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/testdata/flow_control_integration_v2/raft_snapshot_v2_encoding_apply_to_elastic:1:
         
        output didn't match expected:
        @@ -207,11 +207,10 @@
         
           range_id | store_id | total_tracked_tokens  
         -----------+----------+-----------------------
           75       | 1        | 0 B                   
           75       | 2        | 0 B                   
        -  75       | 3        | 0 B                   
           75       | 4        | 0 B                   
           75       | 5        | 0 B                   
         
         
         -- Another view of tokens, using /inspectz-backed vtables.
        --- FAIL: TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2/kvadmission.flow_control.mode=apply_to_elastic (8.58s)
=== RUN   TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2
    --- FAIL: TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2 (21.11s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=23</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFlowControlRaftSnapshotV2.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45893

kv/kvserver.TestFlowControlRaftSnapshotV2 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18580618?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18580618?buildTab=artifacts#/) on master @ [bf4dd7a00b611cbc1d428dd005bd2ec13e78ce1f](https://github.com/cockroachdb/cockroach/commits/bf4dd7a00b611cbc1d428dd005bd2ec13e78ce1f):


```
=== RUN   TestFlowControlRaftSnapshotV2
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/9ded2e1b6e3cd0145900033ce06c11b0/logTestFlowControlRaftSnapshotV21013771654
    test_log_scope.go:76: use -show-logs to present logs inline
    flow_control_integration_test.go:915: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9ded2e1b6e3cd0145900033ce06c11b0/logTestFlowControlRaftSnapshotV21013771654
--- FAIL: TestFlowControlRaftSnapshotV2 (47.06s)
=== RUN   TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2
    --- FAIL: TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2 (23.91s)
=== RUN   TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2/kvadmission.flow_control.mode=apply_to_elastic
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/30655/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/testdata/flow_control_integration_v2/raft_snapshot_v2_encoding_apply_to_elastic:1:
         
        output didn't match expected:
        @@ -258,11 +258,10 @@
         
           range_id | store_id | total_tracked_tokens  
         -----------+----------+-----------------------
           a        | 1        | 0 B                   
           a        | 2        | 0 B                   
        -  a        | 3        | 0 B                   
           a        | 4        | 0 B                   
           a        | 5        | 0 B                   
         
         
         -- Another view of tokens, using /inspectz-backed vtables.
        --- FAIL: TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2/kvadmission.flow_control.mode=apply_to_elastic (11.94s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139229 kv/kvserver: TestFlowControlRaftSnapshotV2 failed [C-bug C-test-failure O-robot T-kv X-duplicate branch-release-25.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFlowControlRaftSnapshotV2.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestFlowControlRaftSnapshotV2 [failed](https://tanzanite.cluster.engflow.com/invocations/default/f996ea8e-bd64-45d4-8748-8adb5264d0a9?testReportRun=19&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [13d8c5191614530b50802f77081f15d2ce2ca705](https://github.com/cockroachdb/cockroach/commits/13d8c5191614530b50802f77081f15d2ce2ca705):


```
=== RUN   TestFlowControlRaftSnapshotV2
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestFlowControlRaftSnapshotV21374101017
    test_log_scope.go:76: use -show-logs to present logs inline
    flow_control_integration_test.go:924: -- test log scope end --
test logs left over in: outputs.zip/logTestFlowControlRaftSnapshotV21374101017
--- FAIL: TestFlowControlRaftSnapshotV2 (45.20s)
=== RUN   TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2
    --- FAIL: TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2 (22.77s)
=== RUN   TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2/kvadmission.flow_control.mode=apply_to_all
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/testdata/flow_control_integration_v2/raft_snapshot_v2_encoding_apply_to_all:1:
         
        output didn't match expected:
        @@ -258,11 +258,10 @@
         
           range_id | store_id | total_tracked_tokens  
         -----------+----------+-----------------------
           a        | 1        | 0 B                   
           a        | 2        | 0 B                   
        -  a        | 3        | 0 B                   
           a        | 4        | 0 B                   
           a        | 5        | 0 B                   
         
         
         -- Another view of tokens, using /inspectz-backed vtables.
        --- FAIL: TestFlowControlRaftSnapshotV2/v2_enabled_when_leader_level=2/kvadmission.flow_control.mode=apply_to_all (11.38s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=19</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139229 kv/kvserver: TestFlowControlRaftSnapshotV2 failed [C-bug C-test-failure O-robot T-kv X-duplicate branch-release-25.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFlowControlRaftSnapshotV2.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/flow_control_integration_test.go
+++ b/pkg/kv/kvserver/flow_control_integration_test.go
@@ -4397,6 +4397,7 @@ func (h *flowControlTestHelper) waitForConnectedStreams(
 	expConnectedStreams, serverIdx int,
 	lvl ...kvflowcontrol.V2EnabledWhenLeaderLevel,
 ) {
+	h.t.Helper()
 	level := h.resolveLevelArgs(lvl...)
 	testutils.SucceedsSoon(h.t, func() error {
 		state, found := h.getInspectHandlesForLevel(serverIdx, level).LookupInspect(rangeID)
@@ -4404,9 +4405,15 @@ func (h *flowControlTestHelper) waitForConnectedStreams(
 			return fmt.Errorf("handle for %s not found", rangeID)
 		}
 		require.True(h.t, found)
-		if len(state.ConnectedStreams) != expConnectedStreams {
-			return fmt.Errorf("expected %d connected streams, got %d",
-				expConnectedStreams, len(state.ConnectedStreams))
+		var connected int
+		for i := range state.ConnectedStreams {
+			if !state.ConnectedStreams[i].Disconnected {
+				connected++
+			}
+		}
+		if connected != expConnectedStreams {
+			return fmt.Errorf("expected %d connected streams, got %d/%d",
+				expConnectedStreams, connected, len(state.ConnectedStreams))
 		}
 		return nil
 	})

--- a/pkg/kv/kvserver/kvflowcontrol/kvflowhandle/testdata/handle_inspect
+++ b/pkg/kv/kvserver/kvflowcontrol/kvflowhandle/testdata/handle_inspect
@@ -26,7 +26,8 @@ echo
       "tracked_deductions": [
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     }
   ]
 }
@@ -65,7 +66,8 @@ echo
         }
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     }
   ]
 }
@@ -104,7 +106,8 @@ echo
         }
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     },
     {
       "stream": {
@@ -128,7 +131,8 @@ echo
         }
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     },
     {
       "stream": {
@@ -152,7 +156,8 @@ echo
         }
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     }
   ]
 }
@@ -191,7 +196,8 @@ echo
         }
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     },
     {
       "stream": {
@@ -215,7 +221,8 @@ echo
         }
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     }
   ]
 }
@@ -238,7 +245,8 @@ echo
       "tracked_deductions": [
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     },
     {
       "stream": {
@@ -254,7 +262,8 @@ echo
       "tracked_deductions": [
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     }
   ]
 }

--- a/pkg/kv/kvserver/kvflowcontrol/kvflowinspectpb/kvflowinspect.proto
+++ b/pkg/kv/kvserver/kvflowcontrol/kvflowinspectpb/kvflowinspect.proto
@@ -79,6 +79,9 @@ message ConnectedStream {
   // before sending, for this connected stream. Only populated on versions >=
   // 24.3.
   int64 total_send_deducted_tokens = 4;
+  // Disconnected is true if the stream has recently disconnected and no longer
+  // actively replicating, i.e. recently left StateReplicate in raft.
+  bool disconnected = 5;
 }
 
 // Stream represents a given kvflowcontrol.Stream and the number of tokens

--- a/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go
+++ b/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go
@@ -1631,6 +1631,7 @@ func (rc *rangeController) InspectRaftMuLocked(ctx context.Context) kvflowinspec
 			}
 			streams = append(streams, kvflowinspectpb.ConnectedStream{
 				Stream:                  rc.opts.SSTokenCounter.InspectStream(rs.stream),
+				Disconnected:            rs.sendStream.mu.connectedState != replicate,
 				TrackedDeductions:       trackedDeductions,
 				TotalEvalDeductedTokens: int64(evalTokens),
 				TotalSendDeductedTokens: int64(sendTokens),

--- a/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/inspect
+++ b/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/inspect
@@ -25,7 +25,8 @@ inspect range_id=1
       "tracked_deductions": [
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     }
   ]
 }
@@ -59,7 +60,8 @@ inspect range_id=1
       "tracked_deductions": [
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     },
     {
       "stream": {
@@ -75,7 +77,8 @@ inspect range_id=1
       "tracked_deductions": [
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     },
     {
       "stream": {
@@ -91,7 +94,8 @@ inspect range_id=1
       "tracked_deductions": [
       ],
       "total_eval_deducted_tokens": "0",
-      "total_send_deducted_tokens": "0"
+      "total_send_deducted_tokens": "0",
+      "disconnected": false
     }
   ]
 }
@@ -153,7 +157,8 @@ inspect range_id=1
         }
       ],
       "total_eval_deducted_tokens": "8388608",
-      "total_send_deducted_tokens": "8388608"
+      "total_send_deducted_tokens": "8388608",
+      "disconnected": false
     },
     {
       "stream": {
@@ -193,7 +198,8 @@ inspect range_id=1
         }
       ],
       "total_eval_deducted_tokens": "8388608",
-      "total_send_deducted_tokens": "8388608"
+      "total_send_deducted_tokens": "8388608",
+      "disconnected": false
     },
     {
       "stream": {
@@ -233,7 +239,8 @@ inspect range_id=1
         }
       ],
       "total_eval_deducted_tokens": "8388608",
-      "total_send_deducted_tokens": "8388608"
+      "total_send_deducted_tokens": "8388608",
+      "disconnected": false
     }
   ]
 }
@@ -292,7 +299,8 @@ inspect range_id=1
         }
       ],
       "total_eval_deducted_tokens": "8388608",
-      "total_send_deducted_tokens": "8388608"
+      "total_send_deducted_tokens": "8388608",
+      "disconnected": false
     },
     {
       "stream": {
@@ -332,7 +340,8 @@ inspect range_id=1
         }
       ],
       "total_eval_deducted_tokens": "8388608",
-      "total_send_deducted_tokens": "8388608"
+      "total_send_deducted_tokens": "8388608",
+      "disconnected": false
     }
   ]
 }
@@ -386,7 +395,8 @@ inspect range_id=1
         }
       ],
       "total_eval_deducted_tokens": "5242880",
-      "total_send_deducted_tokens": "5242880"
+      "total_send_deducted_tokens": "5242880",
+      "disconnected": false
     },
     {
       "stream": {
@@ -410,7 +420,8 @@ inspect range_id=1
         }
       ],
       "total_eval_deducted_tokens": "2097152",
-      "total_send_deducted_tokens": "2097152"
+      "total_send_deducted_tokens": "2097152",
+      "disconnected": false
     }
   ]
 }

